Platform/OpenNtfsDxe/Data.c

```diff
@@ -0,0 +1,566 @@
+   Offset | Size | Description
+   --------------------------------------------------------------------
+   0      | 0.5   | F=Size of the Offset field
+   0.5    | 0.5   | L=Size of the Length field
+   1      | L     | Length of the run
+   1+L    | F     | Offset to the starting LCN of the previous element (signed)
+   --------------------------------------------------------------------
+**/
+EFI_STATUS
+EFIAPI
+ReadRunListElement (
+  IN OUT RUNLIST  *Runlist
+  )
+{
+  UINT8              LengthSize;
+  UINT8              OffsetSize;
+  UINT64             OffsetLcn;
+  UINT8              *Run;
+  ATTR_HEADER_NONRES *Attr;
+  UINT64             BufferSize;
+
+  Run = Runlist->NextDataRun;
+  BufferSize = mFileRecordSize - (Run - Runlist->Attr->BaseMftRecord->FileRecord);
+
+  retry:
+  if (BufferSize == 0) {
+    DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #1) Runlist is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    LengthSize = ((*Run) & 0xF);
+    OffsetSize = ((*Run) >> 4) & 0xF;
+    Run++;
+    BufferSize--;
+  }
+
+  if ((LengthSize > 8) || (OffsetSize > 8) ||
+    ((LengthSize == 0) && (OffsetSize != 0))) {
+    DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #2) Runlist is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  /* End of Runlist: LengthSize == 0, OffsetSize == 0 */
+  if ((LengthSize == 0) && (OffsetSize == 0)) {
+    if ((Runlist->Attr) && (Runlist->Attr->Flags & NTFS_AF_ALST)) {
+      Attr = (ATTR_HEADER_NONRES *) Runlist->Attr->Current;
+      Attr = (ATTR_HEADER_NONRES *) FindAttr (Runlist->Attr, Attr->Type);
+      if (Attr != NULL) {
+
+        if (Attr->NonResFlag == 0) {
+          DEBUG((DEBUG_INFO, "NTFS: $DATA should be non-resident\n"));
+          return EFI_VOLUME_CORRUPTED;
+        }
+
+        Run = (UINT8 *) Attr + Attr->DataRunsOffset;
+        Runlist->CurrentLcn = 0;
+        BufferSize = mFileRecordSize - Attr->DataRunsOffset;
+        goto retry;
+      }
+    }
+    DEBUG((DEBUG_INFO, "NTFS: Run list overflown\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  Runlist->CurrentVcn = Runlist->NextVcn;
+
+  if (BufferSize < LengthSize) {
+    DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #3) Runlist is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    Runlist->NextVcn += ReadField (Run, LengthSize, FALSE);
+    if (Runlist->NextVcn <= Runlist->CurrentVcn) {
+      DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #3.1) Runlist is corrupted.\n"));
+      return EFI_VOLUME_CORRUPTED;
+    }
+
+    Run += LengthSize;
+    BufferSize -= LengthSize;
+  }
+
+  if (BufferSize < OffsetSize) {
+    DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #4) Runlist is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    OffsetLcn = ReadField (Run, OffsetSize, TRUE);
+    Runlist->CurrentLcn += OffsetLcn;
+
+    Run += OffsetSize;
+    BufferSize -= OffsetSize;
+    Runlist->NextDataRun = Run;
+  }
+
+  if (OffsetLcn == 0) {
```

Comment:
Use ternary statements.

Reply:
Ternary statements are often fugly. As this is bool, you can just assign the value immediately.